mir-ci/mir_ci/tests/robot/platforms/zapper/ZapperHid.py

```diff
@@ -0,0 +1,180 @@
+
+    @keyword
+    def type_string(self, string: str):
+        """Type a given string using Zapper HID."""
+        actions = translator.generate_actions_for_typing(string)
+        send_zapper_command("handle_hid_actions", actions)
+
+    def _init_mouse(self) -> None:
+        # TODO: drop this once stylus support is in (ZAP-660)
+        # Reset button state and move to (0, 0) in increments of MOUSE_STEP
+        if self.pressed_buttons is None:
+            self.pressed_buttons = set()
+        if self.mouse_position is None:
+            for _i in range(
+                0, math.ceil(max(*_get_resolution()) / MOUSE_STEP)
+            ):
+                send_zapper_command(
+                    "hid_mouse",
+                    tuple(self.pressed_buttons),
+                    -MOUSE_STEP,
+                    -MOUSE_STEP,
+                    0,
+                )
+            self.mouse_moves = 0
+            self.mouse_position = [0, 0]
+
+    def _move_mouse(self, x, y) -> None:
+        self._init_mouse()
+        self.mouse_position[0] += x
+        self.mouse_position[1] += y
+        send_zapper_command("hid_mouse", tuple(self.pressed_buttons), x, y, 0)
+        self.mouse_moves += 1
+
+    @keyword
+    def press_pointer_button(self, button: str) -> None:
+        """Press the (LEFT|MIDDLE|RIGHT) mouse button"""
+        self._init_mouse()
+        self.pressed_buttons.add(button)
+        send_zapper_command("hid_mouse", tuple(self.pressed_buttons), 0, 0, 0)
+
+    @keyword
+    def release_pointer_button(self, button: str) -> None:
+        """Release the (LEFT|MIDDLE|RIGHT) mouse button"""
+        self._init_mouse()
+        try:
+            self.pressed_buttons.remove(button)
+        except KeyError:
+            pass
+        else:
+            send_zapper_command(
+                "hid_mouse", tuple(self.pressed_buttons), 0, 0, 0
+            )
+
+    @keyword
+    def release_buttons(self) -> None:
+        """Release all pointer buttons"""
+        self._init_mouse()
+        self.pressed_buttons.clear()
+        send_zapper_command("hid_mouse", 0, 0, 0, 0)
+
+    @keyword
+    def click_pointer_button(self, button: str) -> None:
+        """Click the (LEFT|MIDDLE|RIGHT) mouse button"""
+        send_zapper_command("mouse_click", (button,))
+        # To avoid bounce inhibition
+        time.sleep(0.05)
+
+    @keyword
+    def walk_pointer_to_absolute(self, x: int, y: int, step_distance: int, delay: float) -> None:
+        """
+        Walk the virtual pointer to an absolute position within the output, maximum `step_distance`
+        at a time, with `delay` seconds in between.
+        """
+        assert isinstance(x, int) and isinstance(
+            y, int
+        ), "Coordinates must be integers"
+
+        resolution = _get_resolution()
+        assert 0 <= x <= resolution[0], "X coordinate outside of screen"
+        assert 0 <= y <= resolution[1], "Y coordinate outside of screen"
```

Comment:
I wonder if it should be a right-open interval instead. In a 100x100 output, I would expect the in-screen coordinates to be in the range of 0 to 99. OTOH, the current conversions from proportional to absolute coordinates assume that 0..1 is converted to 0..100.